Upgrade react-intl to v3

mcmxcdev opened the pull request:
Based on the comment by @julienben in https://github.com/react-boilerplate/react-boilerplate/pull/2686/#issuecomment-521299449, I decided to upgrade react-intl to its newest version.

Unfortunately I was not able to test if the [polyfills](https://github.com/formatjs/react-intl/blob/master/docs/Upgrade-Guide.md#migrate-to-using-native-intl-apis) are working since I dont have access to neither Safari, Edge or IE on Ubuntu, maybe someone can help out ;)

Commit:
* react-intl upgrade from 2.9.0 to 3.1.9

* new formatjs/intl-relativetimeformat and intl-pluralrules dependencies as react-intl polyfills

* changes in jest.config.js, webpack.base.babel.js, i18n.js and ToggleOption component according to react-intl upgrade guide

* various snapshot updates due to internal react-intl changes

Signed-off-by: mhatvan <markus_hatvan@aon.at>

mcmxcdev commented:
@julienben saw your feedback in https://github.com/react-boilerplate/react-boilerplate/pull/2686#issuecomment-531851190 :
How exactly is this PR broken/incomplete? If its just about the conflicting files thats done fast.
 I was waiting for feedback which I didn't receive yet. Would be happy to finish this ;)

julienben commented:
Hey @mhatvan,

Sorry about that! I just assumed something was broken given that the CI didn't pass. From a quick look at the code, everything seems good and the CI issues seem easy to fix.

I'm assuming the two new dependencies are required by the new version of `react-intl`? I'd like to check the evolution of the build size after this change.

A couple more things:
- Did you check if the documentation needs to be updated? Does it make any reference to now-outdated code? Does it explain how to add a new language and now those instructions should be updated?
- Same for the generators which allow adding a new language. Did you try out `npm run generate language`?

Btw, I have access to a PC with IE and other browsers and will test it out ASAP.

Commit:
Merge branch 'dev' of https://github.com/react-boilerplate/react-boilerplate into chore/upgrade-react-intl-to-v3

Commit:
Upgrade react-intl to v3 (#2721)

* react-intl upgrade from 3.1.9 to 3.3.0

* switch from intl-pluralrules to @formatjs/intl-pluralrules

* update i18n.js and i18n.md

Signed-off-by: mhatvan <markus_hatvan@aon.at>

mcmxcdev commented:
@julienben In my newest commit, I resolved merge conflicts and updated the new dependency imports since the documentation of the actual [upgrade guide](https://github.com/formatjs/react-intl/blob/master/docs/Upgrade-Guide.md) itself changed.
> I'm assuming the two new dependencies are required by the new version of `react-intl`? I'd like to check the evolution of the build size after this change.
>
The new dependencies are necessary due to breaking API changes -> `addLocaleDate` has been 
removed from the `react-intl` dependency. You can read about it here: https://github.com/formatjs/react-intl/blob/master/docs/Upgrade-Guide.md#breaking-api-changes

> Did you check if the documentation needs to be updated? Does it make any reference to now-outdated code? Does it explain how to add a new language and now those instructions should be updated?
>
Good point, I updated the `i18n.md` accordingly.

> Same for the generators which allow adding a new language. Did you try out npm run generate language?
>
Still works as expected, verified it with added fr locale.

> Btw, I have access to a PC with IE and other browsers and will test it out ASAP.
>
That would be great, so that we can get this merged.

vineyardbovines reviewed (approved):
LGTM on my end! Thanks @mhatvan

julienben commented on app/i18n.js:
```diff
@@ -7,16 +8,19 @@
+if (!Intl.RelativeTimeFormat) {
+  require('@formatjs/intl-relativetimeformat/polyfill');
+  require('@formatjs/intl-relativetimeformat/dist/locale-data/de'); // Add locale data for de
+}
```
The sample app has both `de` and `en` so I believe you're missing a couple lines above:

```js
if (!Intl.PluralRules) {
  require('@formatjs/intl-pluralrules/polyfill');
  require('@formatjs/intl-pluralrules/dist/locale-data/de'); // Add locale data for de
  require('@formatjs/intl-pluralrules/dist/locale-data/en'); // Add locale data for en
}

if (!Intl.RelativeTimeFormat) {
  require('@formatjs/intl-relativetimeformat/polyfill');
  require('@formatjs/intl-relativetimeformat/dist/locale-data/de'); // Add locale data for de
  require('@formatjs/intl-relativetimeformat/dist/locale-data/en'); // Add locale data for en
}
```

No?

mcmxcdev replied:
You can read [here](https://www.npmjs.com/package/@formatjs/intl-relativetimeformat#usage) that the en locale is the default locale, the example shows only the de locale being imported explicitly.

julienben replied:
I don't see any mention of `en` being the default locale at this link. IMO, if `@formatjs/intl-relativetimeformat/dist/locale-data/en` didn't need to be included, it wouldn't exist in the `locale-data` folder but it does.

mcmxcdev replied:
I guess I was not imagining it, they removed it from the README.md a while ago: https://github.com/formatjs/formatjs/commit/cda3cf2090abf6a46e293ad049eb7f588f9a66eb#diff-70b3d6472c5178da3042f45cb893ef56

I am updating react-intl and relevant dependencies again, since there were a lot of new version releases in the mean time, I will keep you updated.

julienben replied:
Thanks @mhatvan! Sorry again for the delay on my part!

julienben commented on internals/webpack/webpack.base.babel.js:
```diff
@@ -21,7 +21,8 @@ module.exports = options => ({
         test: /\.jsx?$/, // Transform all .js and .jsx files required somewhere with Babel
-        exclude: /node_modules/,
+        // excluding full node_modules folder except for react-intl related deps, see: https://github.com/formatjs/react-intl/blob/master/docs/Upgrade-Guide.md#webpack
+        exclude: /node_modules\/(?!(react-intl|intl-messageformat|intl-messageformat-parser)\/).*/,
```
IMO, your comment should link to the setup part of the docs instead of the upgrade guide:

https://github.com/formatjs/react-intl/blob/master/docs/Getting-Started.md#esm-build

julienben replied:
Also, did you check if adding those artifacts to the exclude instead of including them specifically mean they do get transpiled correctly? (If you're not sure, I'll find out when checking this upgrade on an IE setup this week.)

mcmxcdev replied:
You are right, changed the link ;)

I checked the output of `npm run build` and the libraries appeared as chunks, transpiled to ES5, but please doublecheck this on IE.

ewief replied:
tested using windows env, as windows have different file paths maybe exclude should be 
`/node_modules[\\,/](?!(react-intl|intl-messageformat|intl-messageformat-parser)[\\,/]).*/`

mcmxcdev replied:
Thank you, I changed it in the newest commit.

julienben reviewed (changes requested):
Hey @mhatvan! Thanks for your hard work on this! Looks great and almost ready to merge. Please see a couple of final comments.

Also, are you sure about the language generator still working? Shouldn't it also add language-specific locale data for the PluralRules and RelativeTimeFormat polyfills?

```js
// In app/i18n.js, when adding for example 'fr':

require('@formatjs/intl-pluralrules/dist/locale-data/fr');

require('@formatjs/intl-relativetimeformat/dist/locale-data/fr');
```

Right now it's adding it using the old format:

```js
const frLocaleData = require('react-intl/locale-data/fr');

addLocaleData(frLocaleData);
```

See `internals/generators/language/index.js` and related HBS templates.

julienben commented:
A couple more things I forgot about:

- Coverage went down from 100% to 99.3%, probably because of polyfills being conditionally required. Can you just add `istanbul ignore` comments to fix that?
- In `app/app.js`, we import `locale-data` as well (+ the generator adds this line for new languages as well). Do you know if this is now redundant because the new polyfill setup takes care of that or not?

(Sorry about this upgrade being so tedious. i18n is always annoying to deal with.)

mcmxcdev commented:
> * In `app/app.js`, we import `locale-data` as well (+ the generator adds this line for new languages as well). Do you know if this is now redundant because the new polyfill setup takes care of that or not?
> 
Hard to get a good overview of the intl stuff, but:
After checking the `react-intl` repo, I saw that they don't have `intl` in their dependencies, and the `intl` bundle languages use `IntlPolyfill.__addLocaleData` internally which is deprecated according to the documentation. 
So I would assume that we could remove the code in `app/app.js`.

Apart from that, all browsers seem to be green for at least basic functionality of `intl`, see: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl#Browser_compatibility

Commit:
Upgrade react-intl to v3 (#2721)

* change add-locale-data.hbs and intl-locale-data.hbs generation

* add istanbul ignore to conditional polyfill imports

* remove intl polyfill in app/app.js and polyfill-intl-locale.hbs

* change documentation url for intl webpack transpilation

Signed-off-by: mhatvan <markus_hatvan@aon.at>

mcmxcdev commented:
@julienben I pushed another commit which updates the language generation, more specifically, it adds the required language polyfills for new languages e.g.: fr

Commit:
Merge branch 'dev' into chore/upgrade-react-intl-to-v3

julienben committed:
Merge branch 'dev' into chore/upgrade-react-intl-to-v3

merqurio commented on app/app.js:
```diff
@@ -69,25 +69,6 @@ if (module.hot) {
-      throw err;
-    });
-} else {
-  render(translationMessages);
```
Hi @mhatvan , thanks for your work.

I tested this PR; removing ther `render()` in line 88 makes the application not load for the first time, although if you make any change in the source code the HMR kicks in and the app renders correctly.  This affects the production build, that doesn't render the application at all. 

_Please, note that I'm working on a fork, but this should have the same results in the template_

Hope this helps with the PR.

mcmxcdev replied:
Hey, thank you for your review!

You are totally right and I am glad you spotted this, I accidentally removed the `render()` when removing the check for `window.Intl`, I re-added it again in the newest commit.

merqurio replied:
Happy to help,

Are you sure we can remove `window.Intl` safely ? Checking [Can I use](https://caniuse.com/#feat=intl-pluralrules) and [Mozilla](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/PluralRules) seems like not all the major browsers support `window.Intl.PluralRules`. Not sure if this applies. 

Bests !

mcmxcdev replied:
The new `@formatjs/intl-pluralrules` dependency should take care of that.

Check up in the react-intl upgrade guide: https://github.com/formatjs/react-intl/blob/master/docs/Upgrade-Guide.md#migrate-to-using-native-intl-apis

And you can read the discussion here: https://github.com/react-boilerplate/react-boilerplate/pull/2721#issuecomment-534292622

Commit:
re-add render() to app.js

Signed-off-by: markushatvan <markus_hatvan@aon.at>

Commit:
change exclude regex in webpack.base.babel.js to work for unix and windows environments

Signed-off-by: markushatvan <markus_hatvan@aon.at>

Commit:
Merge branch 'dev' of https://github.com/react-boilerplate/react-boilerplate into chore/upgrade-react-intl-to-v3

Signed-off-by: mhatvan <markus_hatvan@aon.at>

Commit:
Merge branch 'chore/upgrade-react-intl-to-v3' of github.com:mhatvan/react-boilerplate into chore/upgrade-react-intl-to-v3

nxame commented:
Can we get this through? Please.

julienben committed:
Merge branch 'dev' into chore/upgrade-react-intl-to-v3

julienben reviewed (changes requested):
Sorry it took me so long to review this! I only added one comment regarding `en` as the default locale.

More importantly though, I cloned your fork and tested this branch on a Windows 10 laptop with IE11 and the app was broken. This is what I got in the console:

![8UHVO3c](https://user-images.githubusercontent.com/8948127/69007878-6d0db200-094c-11ea-8bec-7aebf757d322.png)

I'm not sure what the issue is but if you have an idea and would like me to try a fix on PC, let me know. I'll make sure to take care of it faster this time!

Commit:
Merge branch 'dev' of https://github.com/react-boilerplate/react-boilerplate into chore/upgrade-react-intl-to-v3

Commit:
* update react-intl and relevant dependencies to latest version

* add full-icu package to fix locale data missing error in jest tests

* add english locale data polyfill for Intl.Pluralrules and Intl.RelativeTimeFormat

* change language generation to add locale data polyfills after english locale data

Signed-off-by: mhatvan <markus_hatvan@aon.at>

mcmxcdev commented:
> More importantly though, I cloned your fork and tested this branch on a Windows 10 laptop with IE11 and the app was broken. This is what I got in the console:
> 
> ![8UHVO3c](https://user-images.githubusercontent.com/8948127/69007878-6d0db200-094c-11ea-8bec-7aebf757d322.png)
> 
> I'm not sure what the issue is but if you have an idea and would like me to try a fix on PC, let me know. I'll make sure to take care of it faster this time!

@julienben I found out and was able to reproduce this by adding `delete Intl.PluralRules` in `i18n.js` to simulate a kind of faked IE11 mode.
After updating react-intl and related dependencies to the newest versions again, the error you mentioned went away.

I also added the [full-icu](https://www.npmjs.com/package/full-icu) dependency to fix missing locale data errors in jest, see: https://github.com/formatjs/react-intl/issues/1401

Please run the latest commit again locally against your W10 IE11 environment and tell me if there are any remaining issues.

mcmxcdev commented:
@julienben I am aware of the conflicting files, but didn't do any further changes since I want you to decide when you have time and are ready to review this MR.

ividic commented on package.json:
```diff
@@ -59,12 +59,15 @@
     "cross-env": "6.0.3",
     "express": "4.17.1",
     "fontfaceobserver": "2.1.0",
+    "full-icu": "1.3.0",
```
Should this be a devDependency?

mcmxcdev replied:
The install instructions on https://www.npmjs.com/package/full-icu are displaying the commands for putting it in dependencies (no -D flag), but I am open to move it to the devDependencies, if desired.

I hope we can expect some activity from @julienben soon?

cuongdevjs replied:
Hope to check soon. I have a problem with upgrade intl to v3

Can-Sahin commented:
Edit: Sorry I am opening this back as a result of not merging the #2935 if that means anything;)
